In [3]:
import pandas as pd

In [66]:
df = pd.read_csv('financial_data.csv')
df

,Company,Fiscal Year,Total Revenue ($M),Net Income ($M),Total Assets ($M),Total Liabilities ($M),Operating Cash Flow
0,Microsoft,2025,281724.0,101832.0,619003.0,275524.0,136162.0
1,Microsoft,2024,245122.0,88136.0,512163.0,243686.0,118548.0
2,Microsoft,2023,211915.0,72361.0,411976.0,205753.0,87582.0
3,Apple,2025,416161.0,112010.0,359241.0,285508.0,111482.0
4,Apple,2024,391035.0,93736.0,364980.0,308030.0,118254.0
5,Apple,2023,383285.0,96995.0,352583.0,290437.0,110543.0
6,Tesla,2025,94827.0,3855.0,137806.0,54941.0,14747.0
7,Tesla,2024,97690.0,7153.0,122070.0,48390.0,14923.0
8,Tesla,2023,96773.0,14974.0,106618.0,43009.0,13256.0


#### Sort Data

In [68]:
df = df.sort_values(by=['Company', 'Fiscal Year'])
df

,Company,Fiscal Year,Total Revenue ($M),Net Income ($M),Total Assets ($M),Total Liabilities ($M),Operating Cash Flow
5,Apple,2023,383285.0,96995.0,352583.0,290437.0,110543.0
4,Apple,2024,391035.0,93736.0,364980.0,308030.0,118254.0
3,Apple,2025,416161.0,112010.0,359241.0,285508.0,111482.0
2,Microsoft,2023,211915.0,72361.0,411976.0,205753.0,87582.0
1,Microsoft,2024,245122.0,88136.0,512163.0,243686.0,118548.0
0,Microsoft,2025,281724.0,101832.0,619003.0,275524.0,136162.0
8,Tesla,2023,96773.0,14974.0,106618.0,43009.0,13256.0
7,Tesla,2024,97690.0,7153.0,122070.0,48390.0,14923.0
6,Tesla,2025,94827.0,3855.0,137806.0,54941.0,14747.0


#### Growth Calculation 

In [70]:
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue ($M)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income ($M)'].pct_change() * 100
df['Asset Growth (%)'] = df.groupby('Company')['Total Assets ($M)'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby('Company')['Operating Cash Flow'].pct_change() * 100

df

,Company,Fiscal Year,Total Revenue ($M),Net Income ($M),Total Assets ($M),Total Liabilities ($M),Operating Cash Flow,Revenue Growth (%),Net Income Growth (%),Asset Growth (%),Cash Flow Growth (%)
5,Apple,2023,383285.0,96995.0,352583.0,290437.0,110543.0,NaN,NaN,NaN,NaN
4,Apple,2024,391035.0,93736.0,364980.0,308030.0,118254.0,2.021994,-3.359967,3.516052,6.975566
3,Apple,2025,416161.0,112010.0,359241.0,285508.0,111482.0,6.425512,19.495178,-1.572415,-5.726656
2,Microsoft,2023,211915.0,72361.0,411976.0,205753.0,87582.0,NaN,NaN,NaN,NaN
1,Microsoft,2024,245122.0,88136.0,512163.0,243686.0,118548.0,15.669962,21.800417,24.318650,35.356580
0,Microsoft,2025,281724.0,101832.0,619003.0,275524.0,136162.0,14.932156,15.539621,20.860546,14.858117
8,Tesla,2023,96773.0,14974.0,106618.0,43009.0,13256.0,NaN,NaN,NaN,NaN
7,Tesla,2024,97690.0,7153.0,122070.0,48390.0,14923.0,0.947578,-52.230533,14.492862,12.575438
6,Tesla,2025,94827.0,3855.0,137806.0,54941.0,14747.0,-2.930699,-46.106529,12.890964,-1.179388


In [72]:
growth_cols = ['Revenue Growth (%)', 'Net Income Growth (%)']
df[growth_cols] = df[growth_cols].fillna(0)
print("📊 Financial Data with Growth Metrics:\n")
print(df)

📊 Financial Data with Growth Metrics:

     Company  Fiscal Year  Total Revenue ($M)  Net Income ($M)  \
5      Apple         2023            383285.0          96995.0   
4      Apple         2024            391035.0          93736.0   
3      Apple         2025            416161.0         112010.0   
2  Microsoft         2023            211915.0          72361.0   
1  Microsoft         2024            245122.0          88136.0   
0  Microsoft         2025            281724.0         101832.0   
8      Tesla         2023             96773.0          14974.0   
7      Tesla         2024             97690.0           7153.0   
6      Tesla         2025             94827.0           3855.0   

   Total Assets ($M)  Total Liabilities ($M)  Operating Cash Flow  \
5           352583.0                290437.0             110543.0   
4           364980.0                308030.0             118254.0   
3           359241.0                285508.0             111482.0   
2           411976.0    

#### Profit Margin

In [74]:
df['Profit Margin (%)'] = (df['Net Income ($M)'] / df['Total Revenue ($M)']) * 100
df


,Company,Fiscal Year,Total Revenue ($M),Net Income ($M),Total Assets ($M),Total Liabilities ($M),Operating Cash Flow,Revenue Growth (%),Net Income Growth (%),Asset Growth (%),Cash Flow Growth (%),Profit Margin (%)
5,Apple,2023,383285.0,96995.0,352583.0,290437.0,110543.0,0.000000,0.000000,NaN,NaN,25.306234
4,Apple,2024,391035.0,93736.0,364980.0,308030.0,118254.0,2.021994,-3.359967,3.516052,6.975566,23.971256
3,Apple,2025,416161.0,112010.0,359241.0,285508.0,111482.0,6.425512,19.495178,-1.572415,-5.726656,26.915064
2,Microsoft,2023,211915.0,72361.0,411976.0,205753.0,87582.0,0.000000,0.000000,NaN,NaN,34.146238
1,Microsoft,2024,245122.0,88136.0,512163.0,243686.0,118548.0,15.669962,21.800417,24.318650,35.356580,35.955973
0,Microsoft,2025,281724.0,101832.0,619003.0,275524.0,136162.0,14.932156,15.539621,20.860546,14.858117,36.146015
8,Tesla,2023,96773.0,14974.0,106618.0,43009.0,13256.0,0.000000,0.000000,NaN,NaN,15.473324
7,Tesla,2024,97690.0,7153.0,122070.0,48390.0,14923.0,0.947578,-52.230533,14.492862,12.575438,7.322141
6,Tesla,2025,94827.0,3855.0,137806.0,54941.0,14747.0,-2.930699,-46.106529,12.890964,-1.179388,4.065298


#### Company-wise Summary

In [76]:
df.groupby('Company').mean(numeric_only=True)

,Fiscal Year,Total Revenue ($M),Net Income ($M),Total Assets ($M),Total Liabilities ($M),Operating Cash Flow,Revenue Growth (%),Net Income Growth (%),Asset Growth (%),Cash Flow Growth (%),Profit Margin (%)
Company,,,,,,,,,,,
Apple,2024.0,396827.000000,100913.666667,358934.666667,294658.333333,113426.333333,2.815835,5.378404,0.971818,0.624455,25.397518
Microsoft,2024.0,246253.666667,87443.000000,514380.666667,241654.333333,114097.333333,10.200706,12.446679,22.589598,25.107348,35.416075
Tesla,2024.0,96430.000000,8660.666667,122164.666667,48780.000000,14308.666667,-0.661040,-32.779021,13.691913,5.698025,8.953588


In [78]:
summary = df.groupby('Company').agg({
    'Revenue Growth (%)': 'mean',
    'Net Income Growth (%)': 'mean'
}).reset_index()


print("\nYear-over-Year Average Growth Rates (%):")
print(summary)


Year-over-Year Average Growth Rates (%):
     Company  Revenue Growth (%)  Net Income Growth (%)
0      Apple            2.815835               5.378404
1  Microsoft           10.200706              12.446679
2      Tesla           -0.661040             -32.779021


### 📊 Financial Data Analysis Summary
🔍 Overview

This analysis evaluates the financial performance of Microsoft, Apple, and Tesla over the last three fiscal years using key financial metrics such as revenue, net income, assets, liabilities, and operating cash flow. Growth rates and profitability ratios were calculated to identify trends and insights relevant for AI-driven financial analysis.

🏢 Company-wise Insights
Microsoft
Microsoft demonstrates strong and consistent revenue growth (~15% YoY in recent years).

Net income has shown steady growth, indicating strong profitability.
Significant growth in assets reflects business expansion and investment capacity.
Operating cash flow has increased consistently, highlighting excellent liquidity and operational efficiency.
Profit margins remain high (~34–36%), indicating a highly profitable and scalable business model.

👉 Conclusion:
Microsoft is financially robust with strong growth, high profitability, and efficient cash generation, making it a stable and high-performing company.

Apple
Apple shows moderate but stable revenue growth (~2%–6%)

Net income declined slightly in 2024 but recovered strongly in 2025, indicating resilience.
Asset growth is relatively stable, with slight fluctuations.
Operating cash flow remains strong but shows minor volatility.
Profit margins are consistently healthy (~24–27%).

👉 Conclusion:
Apple maintains a strong financial position with stable growth and profitability, though slight fluctuations suggest sensitivity to costs and external factors.

Tesla
Tesla’s revenue growth is stagnant to slightly declining in recent years.
Net income shows a sharp decline (-52% and -46%), indicating significant profitability challenges.
Assets are increasing, suggesting continued investment and expansion.
Operating cash flow is relatively stable but not growing significantly.
Profit margins have dropped drastically (~15% → ~4%), indicating margin pressure.

👉 Conclusion:
Tesla is facing declining profitability despite asset growth, suggesting rising costs, pricing pressure, or operational inefficiencies.

📈 Key Comparative Insights
Microsoft leads in profitability and consistent growth.
Apple shows stability and resilience with strong margins.
Tesla shows declining profitability, making it comparatively riskier.